### DISCUSSION TITLE
ESP32 with ST7789 images from HttpRequests

### DISCUSSION THREAD

**Discussion by Qyther** [Bodmer/TFT_eSPI#2539] [Q&A - General]
Hi, I'm working on a project that requires me to transfer an image from a node server to my **ESP32**. Seeing as how I don't have enough heap to store a whole image, instead I'm sending an initiate/download request to the server, which will download the image on the computer. Then I return a token to the ESP32 which it uses to receive small parts of the image (256 to 1024 or so pixels at a time, I am able to set it to what I want). Then I use **ArduinoJson** to be able to read it, but it outputs it in a **JsonArray**. I have to map it to a small buffer, and then use pushImage() to draw it. This is, however, terribly slow. The loop takes **~30-40 milliseconds per 1024 pixels**. For an image that is 128x128 it will take about **500-600ms** just to do this.

Here is my code:

```
int height = ceil(loadSize / w);
for (int i = 0; i < w * h; i += loadSize) {
  JsonArray loadedImage = readImage(token, i, loadSize)["data"];
  uint16_t imageS[loadSize] = { 0 };
  copyArray(loadedImage, imageS, loadSize);
  int y = floor(i / w);
  tft.pushImage(posx, y + posy, w, min(height, h - y) /*prevent overflow*/, imageS);
}
```

Now my question is: **how could I speed this up besides something like saving it to an SD first and then pushing it? By that I mean that the whole process takes shorter.** Other solutions or recommendations outside of arduino code are welcome! Thanks in advance.

(Edited because I managed to speed it up a little with ArduinoJson, but I would rather it pushes the image whole.)

**Comment by Bodmer** [comment]
Try the library branch and example here:
https://github.com/SirCooper/TJpg_Decoder/tree/master/examples/http_stream

**Reply by Qyther** [reply]
Thank you for your response and aid! However, it is taking around 2.5 seconds to load the small image. My crude homebrewed solution is doing this in 1.6 seconds for the whole screen. Would you know of any faster options?

**Comment by Bodmer** [comment]
This is the only example I have used, it downloads and copies into LittleFS:
https://github.com/Bodmer/TJpg_Decoder/tree/master/examples/LittleFS/LittleFS_Web_Jpg
